### DISCUSSION TITLE
requiresReaction v.s. keepAlive: what's the difference?

### DISCUSSION THREAD

**Discussion by ygnoh** [mobxjs/mobx#3049] [Q&A]
Hello. These days I'm closely looking into the MobX doc. I have many questions.

Anyway...

AFAIK, `keepAlive` keeps `computed` values being cached even if they are not being observed, which means `computed` is always cached anywhere. But `requiresReaction` ... what is it?

The doc says:
> If you try to read its value outside of the reactive context, in which case it might not be cached, it will cause the computed to throw instead of doing an expensive re-evalution.

That means if I use `requiresReaction` then even if the `computed` is not being observed(outside of the reactive context), it will return `computed`. right?

If yes, what is the difference between `requiresReaction` and `keepAlive`?

Thank you.

**Comment by urugator** [comment]
`keepAlive` - computed can be accessed anywhere, the cache is permanent, it does not depend on the presence of `reaction/autorun/observer`.
`requiresReaction` - computed can be accessed only inside `reaction/autorun/observer`, the cache behavior does not change - result is cached only as long as there is at least one `reaction/autorun/observer` observing. Observers can still come and go overtime and so does the cached result. It may reveal accidentally missing `observer`. There is also a global version of this flag.
